### ISSUE TITLE
Solution doesn't scale well with number of users

### ISSUE THREAD

**Issue by pyohannes** [fizban-of-ragnarok/go-gma-server#1]
Due to the usage of a [global lock](https://github.com/fizban-of-ragnarok/go-gma-server/blob/34d346baf9bdf6462bf2b67b16527ba213c6146a/mapservice/mapservice.go#L263), the solution does not scale well with the number of users:
```go
func (c *MapClient) SendToOthers(values ...string) {
	c.Service.lock.RLock()
	for aClientAddr, aClient := range c.Service.Clients {
		if aClientAddr != c.ClientAddr {
			aClient.Send(values...)
		}
	}
	c.Service.lock.RUnlock()
}
```

This blocks the whole `Run` loop, and the more clients are registered, the longer it takes. This pattern with the global lock is not only used at this point, but in many other instances.

It would be far better to handle the sending in the client instead of the service map. A client can have a buffered channel of messages to sent, and a go routine that receives messages from the channel and send them. This would remove the need for global locks on sending functions and make the application much more efficient: clients could send concurrently, whereas now only one client can send a message at a time.

**Comment by MadScienceZone** [comment]
Yes, that's a very valid point. This was a bit of a quick effort to get something functional but there are a couple of places where some refactoring for scalability and efficiency would be recommended.

**Comment by MadScienceZone** [comment]
I have a refactored solution pushed to the refactor_io branch which addresses this issue. Critical paths around shared resources are reduced by moving the client data paths to a buffered channel with a background goroutine to feed data from those to the client sockets without blocking anything else in the process. In places where we got a read-lock to a central list of clients while we performed a write out to each of those clients, this has been replaced with a read-lock-protected access just long enough to get a list of clients, then the lock is released while the code goes on to send data to the channels for those clients, allowing other concurrent operations to take place.
Doubtless there are more opportunities to refactor the code (this was, after all a bit of a quick effort implementation) but this should help the issue reported here considerably.
I'll merge this change to the main branch after testing it more.
